include/cutlass/gemm/collective/xe_mma_w8a8.hpp

```diff
@@ -1,296 +1,332 @@
+
+  // Host side kernel arguments
+  struct Arguments {
+    ElementA const* ptr_A;
+    StrideA dA;
+    ElementB const* ptr_B;
+    StrideB dB;
+  };
+
+  struct Params {
+    Copy_A tiled_copy_a;
+    Copy_B tiled_copy_b;
+  };
+
+  //
+  // Methods
+  //
+
+  CollectiveMma() = default;
+
+  template <class ProblemShape>
+  static constexpr Params
+  to_underlying_arguments(ProblemShape const& problem_shape, Arguments const& args, void* workspace) {
+    (void) workspace;
+
+    auto [M,N,K,L] = problem_shape;
+
+    auto mA_mkl = make_tensor(make_gmem_ptr(args.ptr_A), make_layout(make_shape(M, K, L), args.dA));
+    auto mB_nkl = make_tensor(make_gmem_ptr(args.ptr_B), make_layout(make_shape(N, K, L), args.dB));
+    Copy_A tiled_copy_a{Copy_A{}.with(mA_mkl)};
+    Copy_B tiled_copy_b{Copy_B{}.with(mB_nkl)};
+    
+    return Params{tiled_copy_a, tiled_copy_b};
+  }
+
+  template <class EngineIn,
+      class EngineOut,
+      class LayoutIn,
+      class LayoutOut,
+      class... Ts>
+  CUTLASS_DEVICE
+  void convert_FP8_to_FP16(
+          Tensor<EngineIn, LayoutIn> const& in,
+          Tensor<EngineOut, LayoutOut>& out) {
+
+    static_assert(is_rmem<EngineIn>::value, "Input tensor for A conversion must come from registers");
+    static_assert(is_rmem<EngineOut>::value, "Output tensor for A conversion must come from registers");
+    static_assert(cosize_v<LayoutIn> == cosize_v<LayoutOut>);
+    static_assert(size_v<LayoutIn> == cosize_v<LayoutIn>);
+    static_assert(size_v<LayoutOut> == cosize_v<LayoutOut>);
+
+    using SrcType = typename EngineIn::value_type;
+    using DstType = typename EngineOut::value_type;
+
+    static_assert(std::is_same_v<SrcType, uint8_t>, "Expected fp8 (E4M3) input as uint8_t");
+    static_assert(std::is_same_v<DstType, half_t>, "Expected fp16 output as half_t");
+
+    auto const& src = in(_, _, _);
+    auto const& dst = out(_, _, _);
+
+    SrcType const* pSrc = src.data();
+    DstType* pDst = dst.data();
+
+    constexpr int num_elements = decltype(size(src))::value;
+    // TODO(Codeplay): Move conversion to NumericArrayConverter
+    if constexpr (std::is_same_v<ElementA, float_e5m2_t>) {
+      // Using something as simple as the following code surprisingly
+      // leads to poor performance.
+      // CUTLASS_PRAGMA_UNROLL
+      // for (int i = 0; i < num_elements; i++) {
+      //   reinterpret_cast<uint16_t*>(pDst)[i] = (static_cast<uint16_t>((pSrc[i]))) << 8;
+      // }
+      // The root-cause is unknown, but private memory use is seen in this case.
+      using SrcArray = cutlass::Array<uint8_t, num_elements>;
+      using DstArray = cutlass::Array<uint16_t, num_elements>;
+      SrcArray const* pSrcArr = reinterpret_cast<SrcArray const*>(pSrc);
+      DstArray* pDstArr = reinterpret_cast<DstArray*>(pDst);
+      E5M2_to_FP16<num_elements>(*pSrcArr, *pDstArr);
+    } else {
+      // E4M3 -> FP16 conversion
+      constexpr int chunk_size = 16;

```

Comment:
Hi @min-jean-cho, I've renamed `vec_size` to `chunk_size` since the vectors being created are local to each work-item, and thus map onto the same SIMD lane of an EU. I haven't modified the rest of the E4M3->FP16 implementation.

Please let me know if that's okay with you. Thanks!

Reply:
Sure, thanks for adding e5m2 convert!